node/network/availability-distribution/src/lib.rs

```diff
@@ -14,1228 +14,116 @@
+		Context: SubsystemContext<Message = AvailabilityDistributionMessage> + Sync + Send,
 	{
-		let mut state = ProtocolState {
-			peer_views: HashMap::new(),
-			view: Default::default(),
-			live_under: HashMap::new(),
-			per_relay_parent: HashMap::new(),
-			per_candidate: HashMap::new(),
-		};
-
-		self.run_inner(ctx, &mut state).await
-	}
-
-	/// Start processing work.
-	#[tracing::instrument(skip(self, ctx), fields(subsystem = LOG_TARGET))]
-	async fn run_inner<Context>(self, mut ctx: Context, state: &mut ProtocolState) -> Result<()>
-	where
-		Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
-	{
-		// work: process incoming messages from the overseer.
+		let mut requester = Requester::new(self.keystore.clone()).fuse();
 		loop {
-			let message = ctx
-				.recv()
-				.await
-				.map_err(|e| Error::IncomingMessageChannel(e))?;
-			match message {
-				FromOverseer::Communication {
-					msg: AvailabilityDistributionMessage::NetworkBridgeUpdateV1(event),
-				} => {
-					if let Err(e) = handle_network_msg(
-						&mut ctx,
-						&self.keystore.clone(),
-						state,
-						&self.metrics,
-						event,
-					)
-					.await
-					{
-						tracing::warn!(
-							target: LOG_TARGET,
-							err = ?e,
-							"Failed to handle incoming network messages",
-						);
-					}
+			let action = {
+				let mut subsystem_next = ctx.recv().fuse();
+				futures::select! {
+					subsystem_msg = subsystem_next => Either::Left(subsystem_msg),
+					from_task = requester.next() => Either::Right(from_task),
 				}
-				FromOverseer::Communication {
-					msg: AvailabilityDistributionMessage::AvailabilityFetchingRequest(_),
-				} => {
-					// TODO: Implement issue 2306:
-					tracing::warn!(
-						target: LOG_TARGET,
-						"To be implemented, see: https://github.com/paritytech/polkadot/issues/2306 !",
-					);
+			};
+
+			// Handle task messages sending:
+			let message = match action {
+				Either::Left(subsystem_msg) => {
+					subsystem_msg.map_err(|e| Error::IncomingMessageChannel(e))?
 				}
-				FromOverseer::Signal(OverseerSignal::ActiveLeaves(ActiveLeavesUpdate {
-					activated: _,
-					deactivated: _,
-				})) => {
-					// handled at view change
+				Either::Right(from_task) => {
+					let from_task = from_task.ok_or(Error::RequesterExhausted)??;
+					ctx.send_message(from_task).await;
+					continue;
+				}
+			};
+			match message {
+				FromOverseer::Signal(OverseerSignal::ActiveLeaves(update)) => {
+					// Update the relay chain heads we are fetching our pieces for:
+					requester
+						.get_mut()
+						.update_fetching_heads(&mut ctx, update)
+						.await?;
 				}
 				FromOverseer::Signal(OverseerSignal::BlockFinalized(..)) => {}
 				FromOverseer::Signal(OverseerSignal::Conclude) => {
 					return Ok(());
 				}
+				FromOverseer::Communication {
+					msg: AvailabilityDistributionMessage::AvailabilityFetchingRequest(req),
+				} => {
+					answer_request(&mut ctx, req).await?
```

Comment:
This is one example, but probably a good place to start the dicussion - which errors should be fatal? IMO, failure to get a chunk from the av-store shouldn't take down the entire subsystem.

Reply:
Hmm. I assumed that such internal errors are pretty fatal. What errors coming from the av-store would not be? And what should we do instead?

Reply:
There are potentially spurious disk errors which should not take down the node. I'd suggest `warn!`ing and continuing.

Reply:
I was under the impression, that the overseer restarts failing subsystems. During testing, it looked like I am not even able to bring it down with panics.

Reply:
No, the overseer doesn't restart subystems. Any subsystem going down is fatal. We can't restart subsystems because we'd have a hard time recovering the state of the previous subsystem and starting with a fresh state could be considered "poisoned".